A failed retry with ITestContext  will lose the ITestContext

### TestNG Version
> Note: 7.4.0

### Expected behavior
The example will be executed three times

异常抛出：
java.lang.AssertionError: 
Expected :1
Actual   :2

### Actual behavior
`java.lang.IllegalArgumentException: wrong number of arguments`

Found that the ITestContext context parameter is missing during debugging

### Is the issue reproducible on runner?
all
- [ ] Shell
- [x] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [x] IntelliJ
- [ ] NetBeans


### Test case sample
> Please, share the test case (as small as possible) which shows the issue

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="test" parallel="false">
    <parameter name="id" value="1111"/>
    <parameter name="name" value="qa"/>
    <parameter name="age" value="30"/>

    <listeners>
        <!--rerun after failure-->
        <listener class-name="com.xxxx.xxxx.FailedRetryListener"/>
    </listeners>

    <test verbose="1" name="case1" thread-count="1">
        <classes>
            <class name="com.xxxx.xxx.xxxTest">
                <methods>
                    <include name="create"/>
                </methods>
            </class>
        </classes>
    </test>
</suite>
```

```java
@Parameters({"id", "name", "age"})
@Test(retryAnalyzer= FailedRetry.class)
public void create(String id, String name, String age, ITestContext context){
    Assert.assertEquels(1,2)
}
```

```java
public class FailedRetry implements IRetryAnalyzer {
    private int retryCount = 1;
    private static final int maxRetryCount = 3;

    @Override
    public boolean retry(ITestResult iTestResult) {
        //If an exception is thrown, the failure case will be rerun. If it is an assertion error, try again
        if (iTestResult.getThrowable() instanceof AssertionError && retryCount % maxRetryCount != 0) {
            retryCount++;
            return true;
        } else {
            retryCount = 1;
            return false;
        }
    }
}
```

```java
public class FailedRetryListener implements IAnnotationTransformer {

    @Override
    public void transform(ITestAnnotation iTestAnnotation, Class aClass,
                          Constructor constructor, Method method) {
        {
            IRetryAnalyzer retry = null;
            try {
                retry = iTestAnnotation.getRetryAnalyzerClass().newInstance();
            } catch (InstantiationException | IllegalAccessException e) {
                e.printStackTrace();
            }
            if (retry == null) {
                iTestAnnotation.setRetryAnalyzer(FailedRetry.class);
            }
        }
    }
}
```

> ### TestNG Version
> > Note: 7.4.0
> 
> ### Expected behavior
> The example will be executed three times
> 
> 异常抛出： java.lang.AssertionError: Expected :1 Actual :2
> 
> ### Actual behavior
> `java.lang.IllegalArgumentException: wrong number of arguments`
> 
> Found that the ITestContext context parameter is missing during debugging
> 
> ### Is the issue reproducible on runner?
> all
> 
> * [ ]  Shell
> * [√] Maven
> * [ ]  Gradle
> * [ ]  Ant
> * [ ]  Eclipse
> * [√] IntelliJ
> * [ ]  NetBeans
> 
> ### Test case sample
> > Please, share the test case (as small as possible) which shows the issue
> 
> `
> 
> ```
> <listeners>
>     <!--rerun after failure-->
>     <listener class-name="com.xxxx.xxxx.FailedRetryListener"/>
> </listeners>
> 
> <test verbose="1" name="case1" thread-count="1">
>     <classes>
>         <class name="com.xxxx.xxx.xxxTest">
>             <methods>
>                 <include name="create"/>
>             </methods>
>         </class>
>     </classes>
> </test>
> ```
> 
> `
> 
> `@Parameters({"id", "name", "age"}) @Test(retryAnalyzer= FailedRetry.class) public void create(String id, String name, String age, ITestContext context){ Assert.assertEquels(1,2) }`
> 
> `public class FailedRetry implements IRetryAnalyzer { private int retryCount = 1; private static final int maxRetryCount = 3;
> 
> ```
> @Override
> public boolean retry(ITestResult iTestResult) {
>     //If an exception is thrown, the failure case will be rerun. If it is an assertion error, try again
>     if (iTestResult.getThrowable() instanceof AssertionError && retryCount % maxRetryCount != 0) {
>         retryCount++;
>         return true;
>     } else {
>         retryCount = 1;
>         return false;
>     }
> }
> ```
> 
> } `
> 
> `public class FailedRetryListener implements IAnnotationTransformer {
> 
> ```
> @Override
> public void transform(ITestAnnotation iTestAnnotation, Class aClass,
>                       Constructor constructor, Method method) {
>     {
>         IRetryAnalyzer retry = null;
>         try {
>             retry = iTestAnnotation.getRetryAnalyzerClass().newInstance();
>         } catch (InstantiationException | IllegalAccessException e) {
>             e.printStackTrace();
>         }
>         if (retry == null) {
>             iTestAnnotation.setRetryAnalyzer(FailedRetry.class);
>         }
>     }
> }
> ```
> 
> }`

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-2669: A failed retry with ITestContext will lose the ITestContext. (Nan Liang)
 Fixed: GITHUB-2643: assertEquals(Set,Set) now ignores ordering as it did before. (Elis Edlund)
 Fixed: GITHUB-2653: Assert methods requires casting since TestNg 7.0 for mixed boxed and unboxed primitives in assertEquals.
 Fixed: GITHUB-2229: Restore @BeforeGroups and @AfterGroups Annotations functionality (Krishnan Mahadevan)

--- a/testng-core/src/main/java/org/testng/internal/invokers/TestInvoker.java
+++ b/testng-core/src/main/java/org/testng/internal/invokers/TestInvoker.java
@@ -211,15 +211,7 @@ class TestInvoker extends BaseInvoker implements ITestInvoker {
                 "Encountered problems when gathering parameter values for [%s]. Root cause: ", m);
         throw new DataProviderInvocationException(msg, cause);
       }
-      Object[] parameterValues =
-          Parameters.getParametersFromIndex(
-              Objects.requireNonNull(bag.parameterHolder).parameters,
-              arguments.getParametersIndex());
-      if (bag.parameterHolder.origin == ParameterHolder.ParameterOrigin.NATIVE
-          || bag.parameterHolder.origin == ParameterHolder.ParameterOrigin.ORIGIN_DATA_PROVIDER) {
-        parameterValues = arguments.getParameterValues();
-      }
-
+      Object[] parameterValues = arguments.getParameterValues();
       TestMethodArguments tma =
           new TestMethodArguments.Builder()
               .usingArguments(arguments)

--- a/testng-core/src/test/java/test/SimpleBaseTest.java
+++ b/testng-core/src/test/java/test/SimpleBaseTest.java
@@ -155,6 +155,12 @@ public class SimpleBaseTest {
     return suite;
   }
 
+  protected static XmlSuite createXmlSuite(String suiteName, Map<String, String> params) {
+    XmlSuite suite = createXmlSuite(suiteName);
+    suite.setParameters(params);
+    return suite;
+  }
+
   protected static XmlTest createXmlTestWithPackages(
       XmlSuite suite, String name, String... packageName) {
     XmlTest result = createXmlTest(suite, name);

--- a/testng-core/src/test/java/test/retryAnalyzer/RetryAnalyzerTest.java
+++ b/testng-core/src/test/java/test/retryAnalyzer/RetryAnalyzerTest.java
@@ -4,10 +4,12 @@ import static org.assertj.core.api.Assertions.assertThat;
 
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.stream.Collectors;
+import org.testng.Assert;
 import org.testng.ITestResult;
 import org.testng.TestListenerAdapter;
 import org.testng.TestNG;
@@ -26,6 +28,7 @@ import test.retryAnalyzer.github1600.Github1600TestSample;
 import test.retryAnalyzer.github1706.DataDrivenSample;
 import test.retryAnalyzer.github1706.NativeInjectionSample;
 import test.retryAnalyzer.github1706.ParameterInjectionSample;
+import test.retryAnalyzer.github2669.RetryTestSample;
 import test.retryAnalyzer.issue1241.GitHub1241Sample;
 import test.retryAnalyzer.issue1538.TestClassSampleWithTestMethodDependencies;
 import test.retryAnalyzer.issue1697.DatadrivenSample;
@@ -238,6 +241,20 @@ public class RetryAnalyzerTest extends SimpleBaseTest {
     TestNG testng = create(xmlsuite);
     testng.run();
     assertThat(test.retryAnalyzer.dataprovider.issue2163.RetryAnalyzer.logs).hasSize(21);
+  }
+
+  @Test(description = "GITHUB-2669")
+  public void testFailedRetryWithParameters() {
+    Map<String, String> params = new HashMap<>();
+    params.put("id", "1111");
+    params.put("name", "qa");
+    params.put("age", "30");
+    XmlSuite suite = createXmlSuite("GITHUB_2669", params);
+
+    createXmlTest(suite, "2669_Test", RetryTestSample.class);
+    TestNG testng = create(suite);
+    testng.run();
+    Assert.assertEquals(RetryTestSample.count, 3);
   }
 
   private ITestResult runAssertions(Set<ITestResult> results, String methodName) {

--- a/testng-core/src/test/java/test/retryAnalyzer/github2669/FailedRetry.java
+++ b/testng-core/src/test/java/test/retryAnalyzer/github2669/FailedRetry.java
@@ -1,0 +1,21 @@
+package test.retryAnalyzer.github2669;
+
+import org.testng.IRetryAnalyzer;
+import org.testng.ITestResult;
+
+public class FailedRetry implements IRetryAnalyzer {
+  public static int retryCount = 1;
+  private static final int maxRetryCount = 3;
+
+  @Override
+  public boolean retry(ITestResult iTestResult) {
+    // If an exception is thrown, the failure case will be rerun. If it is an assertion error, try
+    // again
+    if (iTestResult.getThrowable() instanceof AssertionError && retryCount % maxRetryCount != 0) {
+      retryCount++;
+      return true;
+    }
+    retryCount = 1;
+    return false;
+  }
+}

--- a/testng-core/src/test/java/test/retryAnalyzer/github2669/RetryTestSample.java
+++ b/testng-core/src/test/java/test/retryAnalyzer/github2669/RetryTestSample.java
@@ -1,0 +1,17 @@
+package test.retryAnalyzer.github2669;
+
+import org.testng.Assert;
+import org.testng.ITestContext;
+import org.testng.annotations.Parameters;
+import org.testng.annotations.Test;
+
+public class RetryTestSample {
+  public static int count = 0;
+
+  @Parameters({"id", "name", "age"})
+  @Test(retryAnalyzer = FailedRetry.class)
+  public void create(String id, String name, String age, ITestContext context) {
+    count++;
+    Assert.fail();
+  }
+}

